README.md

```diff
@@ -1,19 +1,122 @@
+```Haskell
+fooM :: DualMonad d r m 
+     => DualNumberVariables d r -> m (DualNumber d r) 
+fooM variables = do { 
+     let x : y : z : _ = vars variables 
+     w <- x * sin y 
+     returnLet $ (atan2 z w + z * w)  -- [note extra w, we *want* to require returnLet up front --awf] 
+   }
+```
+We hope to automate this transfomation soon, automatically generating `fooM` from `foo`, but for now it's a mechanical, if somewhat tedious, manual task.
+
+And then the gradient of `foo` is
+```Haskell
+grad_foo :: HasDelta r => Domain0 r -> (Domain0 r, r) 
+grad_foo ds = 
+   let ((result, _, _, _), value) = 
+         dReverse 1 fooM (ds, V.empty, V.empty, V.empty) 
+   in (result, value) 
+```
+
+As can be verified by computing the gradient at `(1.1,2.2,3.3)`:
+```Haskell
+>>> grad_foo (mkDual [1.1, 2.2, 3.3])
+[7.75, 1.23, 5.52]
+```
+
+## Computing Jacobians.
+Now let's consider a function from 'R^n -> R^m'.  We don't want the gradient, but instead the Jacobian.
+```Haskell
+-- A function that goes from `R^3` to `R^2`. 
+foo :: RealFloat a => (a,a,a) -> (a,a)
+foo (x,y,z) = 
+  let w = x * sin y 
+  in (atan2 z w, z * w)
+```
+TODO: show how the 2x3 Jacobian emerges from here
+
+# Forall shapes and sizes
+
+An additional feature of this library is a type system for tensor shape arithmetic.  Look at this beauty.
+```Haskell
+convMnistTwoS x                                 -- input images, shape (batch_size, c_in, h, w)
+              ker1 bias1                        -- layer1 kernel, shape (c_out, c_in, kh+1, kw+1); and bias, shape (c_out)
+              ker2 bias2                        -- layer2 kernel, shape (c_out, c_out, kh+1, kw+1); and bias, shape (c_out)
+              weigthsDense                      -- dense weights shape (n_hidden, c_out, ((h+kh)/2 + kh)/2, ((w+kw)/2 + kw)/2) 
+              biasesDense                       -- dense biases shape (n_hidden)
+              weigthsReadout biasesReadout      -- readout layer weights (10, n_hidden) and bias (10)
+              =                                 --> output (10, batch_size)
+  do
+  t1 <- convMnistLayerS ker1 (constant x) bias1
+  t2 <- convMnistLayerS ker2 t1 bias2
+  let m1 = mapS reshapeS t2
+      m2 = from2S (transpose2 (fromS2 m1))  -- TODO: add permuation transposeS
+      denseLayer = weigthsDense <>$ m2 + asColumnS biasesDense
+  denseRelu <- reluAct denseLayer
+  returnLet $ weigthsReadout <>$ denseRelu + asColumnS biasesReadout
+```
+But we don't just want the shapes in comments, we want them in the type system:
+```Haskell
+convMnistTwoS
+  :: forall kh kw num_hidden c_out h w c_in batch_size d r m.
+     ( KnownNat kh, KnownNat kw, KnownNat num_hidden, KnownNat c_out, KnownNat h, KnownNat w, KnownNat batch_size
+     , c_in ~ 1 -- greyscale
+     , 1 <= kh  -- kh greater than 1
+     , 1 <= kw  -- kw greater than 1
+     , DualMonad d r m )
+  => OS.Array '[batch_size, c_in, h, w] r
+  -> DualNumber d (OS.Array '[ c_out, c_in , kh + 1, kw + 1 ] r)
```

Comment:
To fit such things into a single line, perhaps the `DualNumber` type constructor could be renamed to `D` (the same as the value constructor of that type). That's what @tomjaguarpaw did in his variant of the implementation. `DualNumber` is less cryptic at the first glance, but more noisy in the long run, as we can see here.